### DISCUSSION TITLE
typesafe references to named queries and result set mappings

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#4483] [Design Proposals]
We should enhance the jpamodelgen to produce things that can be used as typesafe references to `@NamedQuery`s and `@ResultSetMapping`s.

This could in principle be as simple as just a 

```java
public static final String customerByName = "customerByName";
```

for `@NamedQuery(name="customerByName", ....)` but in principle we could parse the query and determine its result type, letting us generate something like:

```java
public static final NamedQuery<Customer> customerByName = new NamedQuery<>("customerByName");
```

I would be happy to work on this if someone who knows how jpamodelgen works could point me in the right direction.

**Comment by gavinking** [comment]
> but in principle we could parse the query and determine its result type

Note that for `@ResultSetMapping` it's even easier to determine the result type.

**Comment by beikov** [comment]
Sounds like a nice addition, but I think there are a few things to consider.

* If we want to use this `NamedQuery` contract we have to define it first, which would be part of hibernate-core, hence users would be required to depend on hibernate-core if they want to use that feature. We can't simply expect users to do that IMO and ergo this should be deactivated by default. How about providing static factories instead? `public static TypedQuery<Customer> customerByName(EntityManager em) {...}`. Heck, we could even try to parse out the parameter names and try to infer the types :)
* Another issue is, that named queries do not necessarily have to be part of an entity. In XML they are listed without entity context. So where should we put the factory method? Maybe we should emit a dedicated class that contains all named queries instead?

Anyway, you'd probably have to analyze annotations in `org.hibernate.jpamodelgen.annotation.AnnotationMetaEntity` to register named queries etc. and do the same in `org.hibernate.jpamodelgen.xml.JpaDescriptorParser` for the XML defined ones. Finally you'd have to emit the special class in `org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor#createMetaModelClasses`

**Comment by gavinking** [comment]
This (and much more) has been implemented.
